### PR TITLE
Add--max-total-trials parameter

### PR DESCRIPTION
**Pull request by ernestum**
Add a `--total-n-trials parameter` to help with parallel tuning

## Description
The MaxTrialsCallback is used to ensure that no more than total-n-trials are created when it is set

## Motivation and Context
When doing parallel tuning it is annoying to divide n-trials by the number of workers. Especially when the number of workers changes over time.

- [ ] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: utils/exp_manager.py [hunk 1]
```diff
@@ -747,7 +751,14 @@ def hyperparameters_optimization(self) -> None:
         )
 
         try:
-            study.optimize(self.objective, n_trials=self.n_trials, n_jobs=self.n_jobs)
+            if self.total_n_trials is not None:
```
Comment (araffin) [inline utils/exp_manager.py:754]: I would prefer the if clause to be only about the callback argument

**Review by araffin — COMMENTED** [review]
Thanks for the PR, please add a test, update the changelog, rebase/merge master and we are good to go ;)

**Comment by araffin** [comment]
@jkterry1 this would solve your issue from https://github.com/DLR-RM/rl-baselines3-zoo/issues/153

**Comment by araffin** [comment]
btw, I cannot edit/push to your fork, so could you merge master into your branch?

**Comment by ernestum** [comment]
> btw, I cannot edit/push to your fork, so could you merge master into your branch?

Sure I just rebased.

**Comment by ernestum** [comment]
I had to add some more lines you maybe you want to review again. I also was not sure how you wanted the if clause to change.

Also see https://github.com/optuna/optuna/issues/3518 for the context/reason of the changes.

**Comment by araffin** [comment]
> I had to add some more lines you maybe you want to review again. I also was not sure how you wanted the if clause to change.

Because of https://github.com/optuna/optuna/issues/3518, my remark doesn't apply anymore ;)

**Comment by ernestum** [comment]
Do you have a clue where the sqlite issue in the failed test might come from @araffin? I already put some time into it but am kind of stuck.

**Comment by araffin** [comment]
I also remember that sqlite may not be thread-safe.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,7 @@
 
 ### New Features
 - Support setting PyTorch's device with thye `--device` flag (@gregwar)
+- Add `--tital-n-trials` parameter to help with distributed optimization. (@ernestum)
```
Comment (nzw0301) [inline CHANGELOG.md:11]:
```suggestion
- Add `--total-n-trials` parameter to help with distributed optimization. (@ernestum)
```

File: utils/exp_manager.py [hunk 3]
```diff
@@ -747,7 +751,26 @@ def hyperparameters_optimization(self) -> None:
...
+                counted_states = [
+                    TrialState.COMPLETE,
+                    TrialState.RUNNING,
+                    TrialState.PRUNED,
+                ]
+                completed_trials = sum(t.state in counted_states for t in study.trials)
```
Comment (nzw0301) [inline utils/exp_manager.py:760]: Optuna has an API to get all trials by trial state:
```suggestion
                completed_trials = len(study.get_trials(state= counted_states))
```

**Comment by ernestum** [comment]
Thanks at @nzw0301 for the valuable feedback!

> I also remember that sqlite may not be thread-safe.

It should be thread-safe but it does not really seem to support concurrent writes and they don't recommend to to use it for distributed optimization:

https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html

So apart from setting up a full SQL server just for the tests, what would you recommend to properly implement the test?

> Does it work locally?

No I (randomly) have the same issues locally. No deterministic way to reproduce it unfortunately.

**Comment by araffin** [comment]
so what we can do is actually single processes, one after another, no?
this would be better than nothing and also deterministic.

**Comment by jkterry1** [comment]
@araffin this would solve my problem, but "n-trials" vs "total-n-trials" are mildly confusing names together in my opinion (though the doc string is good)

**Comment by araffin** [comment]
>  are mildly confusing names together in my opinion

true, do you have better names to suggest?

**Comment by ernestum** [comment]
My only idea would be to make it `--n-additional-trials` and `--n-total-trials` to make the distinction clear. We could still keep `--n-trials` for backwards compatibility.
What do you think? @araffin and @jkterry1 ?

**Comment by araffin** [comment]
> My only idea would be to make it `--n-additional-trials` and `--n-total-trials` to make the distinction clear. We could still keep `--n-trials` for backwards compatibility. What do you think? @araffin and @jkterry1 ?

Maybe keep `--n-trials` and rename the other to `--max-n-trials` or `--max-total-trials`? (`--n-additional-trials` suggest that you add up to some existing ones)

**Comment by ernestum** [comment]
I am fine with `--max-total-trials`. I will rename it to that.

**Comment by ernestum** [comment]
I am ready for a merge if the tests pass after the rename.
Note: I changed the tests to use redis instead of sqlite since it just caused trouble. For that I added a redis service to the CI. I took this idea from the tests of optuna itself: https://github.com/optuna/optuna/blob/master/.github/workflows/tests-storage.yml

File: tests/test_train.py [hunk 4]
```diff
@@ -101,3 +103,47 @@ def test_parallel_train(tmp_path):
 
     return_code = subprocess.call(["python", "train.py"] + args)
     _assert_eq(return_code, 0)
+
+
+def test_multiple_workers(tmp_path):
```
Comment (araffin) [inline tests/test_train.py:108]: could you skip that one if redis is not installed? (that would be the case locally or in docker)

**Review by araffin — DISMISSED** [review]
One small comment, otherwise LGTM, thanks =)

**Review by araffin — APPROVED** [review]

**Pull request merged**